framework/basehandlers.py

```diff
@@ -413,11 +413,11 @@ def split_emails(self, param_name):
   def parse_link(self, param_name):
     link = flask.request.form.get(param_name) or None
     if link:
-      if not link.startswith('http'):
-        link = str('http://' + link)
-      else:
-        link = str(link)
-    return link
+      # see https://www.regextester.com/93901 for url regex
+      url_pattern = (r'\b([\w_-]+(?:(?:\.[\w_-]+)+))'
+                     r'([\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])?\b')
+      match_obj = re.search(url_pattern, str(link))
+    return 'http://' + match_obj.group() if match_obj else None
```

Comment:
Here, you are making every URL use the `http://` scheme.   It would be better to keep the `https://` scheme if the user entered it.    And, in the case where the user entered no scheme or something that we do not support (like `ftp`, `file`, or `about`), then use `http://`.

Reply:
I modified the regex slightly so that it reserves the ```http://``` or ```https://``` of the URL. And now we concatenate the link with ```http://``` only when it does not start with ```http://``` or ```https://```.